File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/TradeItSDK.java
package it.trade.android.sdk;


import android.content.Context;
import android.util.Log;

import it.trade.android.sdk.exceptions.TradeItSDKConfigurationException;
import it.trade.android.sdk.manager.TradeItLinkedBrokerManager;
import it.trade.android.sdk.model.TradeItLinkedBrokerCache;
import it.trade.model.request.TradeItEnvironment;


public class TradeItSDK {
    private static TradeItSdkInstance instance;


    public static void configure(TradeItConfigurationBuilder configurationBuilder) {
        if (instance == null) {
            instance = new TradeItSdkInstance(configurationBuilder);
        } else {
            Log.w("TradeItSDK", "Warning: TradeItSDK.configure() called multiple times. Ignoring.");
        }
    }

    public static void clearConfig() {
        instance = null;
    }

    public static TradeItLinkedBrokerManager getLinkedBrokerManager() throws TradeItSDKConfigurationException {
        if (instance == null) {
            throw new TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerManager referenced before calling TradeItSDK.configure()!");
        }
        return instance.getLinkedBrokerManager();
    }

    public static TradeItEnvironment getEnvironment() {
        if (instance == null) {
            throw new TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerManager referenced before calling TradeItSDK.configure()!");
        }
        return instance.getEnvironment();
    }

    public static String getApiKey() {
        if (instance == null) {
            throw new TradeItSDKConfigurationException("ERROR: TradeItSDK.apiKey referenced before calling TradeItSDK.configure()!");
        }
        return instance.getApiKey();
    }

    public static TradeItLinkedBrokerCache getLinkedBrokerCache() {
        if (instance == null) {
            throw new TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerCache referenced before calling TradeItSDK.configure()!");
        }
        return instance.getLinkedBrokerCache();
    }

    public static Context getContext() {
        if (instance == null) {
            throw new TradeItSDKConfigurationException("ERROR: TradeItSDK.context referenced before calling TradeItSDK.configure()!");
        }
        return instance.getContext();
    }
}

File: tradeit-android-sdk/src/main/kotlin/it/trade/android/sdk/TradeItSDK.kt
package it.trade.android.sdk


import android.util.Log

import it.trade.android.sdk.exceptions.TradeItSDKConfigurationException
import it.trade.android.sdk.manager.TradeItLinkedBrokerManager
import it.trade.android.sdk.model.TradeItLinkedBrokerCache
import it.trade.model.request.TradeItEnvironment


object TradeItSDK {
    private var instance: TradeItSdkInstance? = null

    @JvmStatic
    val linkedBrokerManager: TradeItLinkedBrokerManager?
        @Throws(TradeItSDKConfigurationException::class)
        get() {
            if (instance == null) {
                throw TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerManager referenced before calling TradeItSDK.configure()!")
            }
            return instance!!.linkedBrokerManager
        }

    val environment: TradeItEnvironment?
        get() {
            if (instance == null) {
                throw TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerManager referenced before calling TradeItSDK.configure()!")
            }
            return instance!!.environment
        }

    val linkedBrokerCache: TradeItLinkedBrokerCache?
        get() {
            if (instance == null) {
                throw TradeItSDKConfigurationException("ERROR: TradeItSDK.linkedBrokerCache referenced before calling TradeItSDK.configure()!")
            }
            return instance!!.linkedBrokerCache
        }


    @JvmStatic
    fun configure(configurationBuilder: TradeItConfigurationBuilder) {
        if (instance == null) {
            instance = TradeItSdkInstance(configurationBuilder)
        } else {
            Log.w("TradeItSDK", "Warning: TradeItSDK.configure() called multiple times. Ignoring.")
        }
    }

    fun clearConfig() {
        instance = null
    }
}
